Internal Load Balancer Controller do NOT remove nodes with empty string in the zone field

08volt opened the pull request:
Fix(ILB): Prevent deletion of nodes with empty zone label

Previously, new nodes were created with all topology labels present, including the zone label. However, recently, labels are being patched/updated on the node resource *after* the resource is created. This can result in nodes having an empty zone field when initially added to an instance group.

The Internal Load Balancer controller was incorrectly removing these nodes with empty zone labels from their assigned instance groups. This occurred because the controller interpreted nodes with empty zones as not belonging to any zone.

This commit fixes the issue by:

- Storing nodes with empty zones during node processing.
- Preventing the controller from removing these nodes during zone processing.

This ensures that new nodes with initially empty zone labels, due to the asynchronous label patching, are not prematurely deleted from their instance groups.

The fix has been verified with a new test which add a node with an empty zone label to an instance group and confirming that it is not deleted after a controller sync.

aojea commented:
/ok-to-test
/assign @cezarygerard

cezarygerard commented:
/retest

mmamczur commented on providers/gce/gce_loadbalancer_internal_test.go:
```diff
@@ -705,6 +705,32 @@ func TestUpdateInternalLoadBalancerNodes(t *testing.T) {
+
+	// Expect load balancer to not have deleted node test-node-1
+	node := &v1.Node{ObjectMeta: metav1.ObjectMeta{Name: nodeName}}
+	exist, err := gce.InstanceExists(context.TODO(), node)
```
I think this test is wrong
it is supposed to test if a node is present in the instance group and then if that node zone is removed that it won't be removed from the group

I see something like
- create fake node with a zone
- create service
- make sure the node is in the instance group
- remove zone from the node
- trigger update somehow (node modification should do it?)
- make sure node is still in the instance group

here you just check if the instance exists (and it will as long as the node exists)

08volt replied:
You are right, working on it

08volt commented:
/retest

08volt commented:
/retest

08volt commented:
/retest

08volt commented:
/retest

08volt commented:
/retest

08volt commented:
@cezarygerard @aojea 
can we merge this?

mmamczur commented:
/lgtm

bowei commented:
Can we add more information in the Git commit message (not just the Github PR):

- When does this condition come up -- basically why would you get an empty Zone in the Node resource -- seems impossible?
- Why not ignoring the missing zone field is safe. Is the information coming from a different place?
- How this change was tested. I assume there is some situation where this behavior was causing a problem -- how did we test that we have fixed the problem.

08volt commented:
> Can we add more information in the Git commit message (not just the Github PR). Use `git commit --amend` to edit the commit message.
> 
> * When does this condition come up -- basically why would you get an empty Zone in the Node resource -- seems impossible?
> * Why not ignoring the missing zone field is safe. Is the information coming from a different place?
> * How this change was tested. I assume there is some situation where this behavior was causing a problem -- how did we test that we have fixed the problem.

Thanks for the input, I updated the commit message with new information. Please let me know if you think it is enough.

bowei commented on providers/gce/gce_loadbalancer_internal.go:
```diff
@@ -700,7 +719,7 @@ func (g *Cloud) ensureInternalInstanceGroups(name string, nodes []*v1.Node) ([]s
 			}
 		} else {
-			igLink, err := g.ensureInternalInstanceGroup(name, zone, nodes)
+			igLink, err := g.ensureInternalInstanceGroup(name, zone, nodes, zonedNodes[""])
```
How do we know that all Nodes with an empty zone belong to the zone being ensured here?

Specifically:

```
igLink, err := g.ensureInternalInstanceGroup(name, zone, nodes, zonedNodes[""])
                                                   ^^^^         ^^^^^^^^^^^^^^
                                                    A            B
```

That B all belong in zone A.

mmamczur replied:
so the logic here is that zonedNodes contain nodes that 'should not be removed from Instance Groups'.

What this means is that if, a node is attached to an IG, but for some reason it's zone assignment disappears, don't remove it from the group.
nodes in `emptyZoneNodes` will not be added, it's just to keep them from being removed.

inside `ensureInternalInstanceGroup` there is this part `removeNodes := gceNodes.Difference(kubeNodes).Difference(emptyZoneNodesNames).List()` - when determining which nodes to remove, skip the empty zones one.

In general we have many issues that start with nodes becoming 'temporarily' unready, or missing some data like zone info. Then LB controllers start removing them and we get outages.

mmamczur commented:
/lgtm
/approve
